xonsh/events.py

```diff
@@ -0,0 +1,101 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.doc("Comes with eggs")
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guaranteed.
+    """
+
+    @classmethod
+    def doc(cls, text):
+        cls.__doc__ = text
+
+    def __call__(self, func):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+
+        A decorator method is added to the handler, validator(). If a validator
+        function is added, it can filter if the handler will be considered. The
+        validator takes the same arguments as the handler. If it returns False,
+        the handler will not called or considered, as if it was not registered
+        at all.
+        """
+        #  Using Pythons "private" munging to minimize hypothetical collisions
+        func.__validator = None
+        self.add(func)
+
+        def validator(vfunc):
+            """
+            Adds a validator function to a handler to limit when it is considered.
+            """
+            func.__validator = vfunc
+        func.validator = validator
+
+        return func
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+        """
+        for handler in self:
+            if handler.__validator is not None and not handler.__validator(*pargs, **kwargs):
+                continue
+            yield handler(*pargs, **kwargs)
+
+    def fire(self, *pargs, **kwargs):
+        """
+        The simplest use case: Calls each handler in turn with the provided
+        arguments and ignore the return values.
+        """
+        for _ in self.calleach(*pargs, **kwargs):
+            pass
+
+    def until_true(self, *pargs, **kwargs):
+        """
+        Calls each handler until one returns something truthy.
+
+        Returns that truthy value.
+        """
+        for rv in self.calleach(*pargs, **kwargs):
+            if rv:
+                return rv
+
+    def until_false(self, *pargs, **kwargs):
+        """
+        Calls each handler until one returns something falsey.
+        """
+        for rv in self.calleach(*pargs, **kwargs):
+            if not rv:
+                return rv
+
+
+class Events:
+    """
+    Container for all events in a system.
+
+    Meant to be a singleton, but doesn't enforce that itself.
+
+    Each event is just an attribute. They're created dynamically on first use.
+    """
+
+    def __getattr__(self, name):
```

Comment:
Ok, here's the thing:

If you have a central event registry, you need to have events be usable before the "owning" module (because such a concept doesn't really exist) is loaded, because otherwise you have an implicit dependency. If the event system is used in the core, what if the handling module is loaded before the "registering" module?

The alternative is to have event objects instantiated and scattered about the code base, such that:

```
from xonsh.exec import on_precommand

@on_precommand
def myhandler(cmd):
    ...
```

I'm not saying either one is wrong, but I will say that anything that causes implicit, untracable dependencies is bad.

(Note that by the time the _user_ sees the event registry, the list of available events will be known, except for xontribs that are loaded by the user later.)